File: CPP/Arrays/D12_MaxCircularSubarraySum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int kadane(const vector<int> &nums)
{
      int maxEndingHere = nums[0], maxSoFar = nums[0];
      for (size_t i = 1; i < nums.size(); i++)
      {
            maxEndingHere = max(nums[i], maxEndingHere + nums[i]);
            maxSoFar = max(maxSoFar, maxEndingHere);
      }
      return maxSoFar;
}

int circularSubarraySum(vector<int> &nums)
{
      int totalSum = 0, currMin = 0, minSum = 0;
      int maxKadane = kadane(nums);

      for (int num : nums)
      {
            totalSum += num;
            currMin = min(num, currMin + num);
            minSum = min(minSum, currMin);
      }

      if (maxKadane < 0)
            return maxKadane;

      return max(maxKadane, totalSum - minSum);
}

int main()
{
      vector<int> arr1 = {8, -8, 9, -9, 10, -11, 12};
      vector<int> arr2 = {10, -3, -4, 7, 6, 5, -4, -1};
      vector<int> arr3 = {-1, 40, -14, 7, 6, 5, -4, -1};

      cout << circularSubarraySum(arr1) << endl;
      cout << circularSubarraySum(arr2) << endl;
      cout << circularSubarraySum(arr3) << endl;
      return 0;
}

File: Python/Arrays/D12_MaxCircularSubarraySum.py
def kadane(nums):
    max_ending_here = max_so_far = nums[0]
    for num in nums[1:]:
        max_ending_here = max(num, max_ending_here + num)
        max_so_far = max(max_so_far, max_ending_here)
    return max_so_far

def circular_subarray_sum(nums):
    total_sum = 0
    curr_min = 0
    min_sum = 0

    max_kadane = kadane(nums)

    for num in nums:
        total_sum += num
        curr_min = min(num, curr_min + num)
        min_sum = min(min_sum, curr_min)

    if max_kadane < 0:
        return max_kadane

    return max(max_kadane, total_sum - min_sum)

arr1 = [8, -8, 9, -9, 10, -11, 12]
arr2 = [10, -3, -4, 7, 6, 5, -4, -1]
arr3 = [-1, 40, -14, 7, 6, 5, -4, -1]

print(circular_subarray_sum(arr1))
print(circular_subarray_sum(arr2))
print(circular_subarray_sum(arr3))